boards/native/native_sim/reboot.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2025 GARDENA GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define _DEFAULT_SOURCE
+
+#include "posix_board_if.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+
+#include <fcntl.h>
+#include <unistd.h>
+
+/* arbitrary sizes */
+#define CMDLINE_BUF_SIZE 4096
+#define ARGV_BUF_SIZE    1024
+
+static const char module[] = "native_sim_reboot";
+
+static ssize_t read_cmdline(char cmdline[], size_t *len)
+{
+	int fd = open("/proc/self/cmdline", O_RDONLY | O_CLOEXEC);
+
+	if (fd < 0) {
+		perror(module);
+		return fd;
+	}
+
+	*len = 0;
+	for (ssize_t read_len = -1; read_len != 0;) {
+		if (*len >= CMDLINE_BUF_SIZE) {
+			fprintf(stderr, "%s: Command line is too long.\n", module);
+			close(fd);
+			return -1;
+		}
+		/* the last byte in cmdline is not used */
+		read_len = read(fd, cmdline + *len, CMDLINE_BUF_SIZE - *len);
+		if (read_len < 0) {
+			perror(module);
+			close(fd);
+			return read_len;
+		}
+		*len += read_len;
+	}
+
+	close(fd);
+
+	/* ensure null termination, in case 0 bytes were read */
+	cmdline[*len] = '\0';
+	return 0;
+}
+
+static int get_args(char cmdline[], size_t len, char *argv[])
+{
+	char *p = cmdline;
+	size_t i = 0;
+
+	while (p < cmdline + len) {
+		if (i + 1 >= ARGV_BUF_SIZE) {
+			fprintf(stderr, "%s: Command line has too many arguments.\n", module);
+			return -1;
+		}
+		argv[i] = p;
+		p += strlen(p) + 1;
+		i++;
+	}
+	argv[i] = NULL;
+	return 0;
+}
+
+static long close_open_fds(void)
+{
+	/* close all open file descriptors except 0-2 */
+	errno = 0;
+
+	long max_fd = sysconf(_SC_OPEN_MAX);
+
+	if (max_fd < 0) {
+		if (errno != 0) {
+			perror(module);
+		} else {
+			fprintf(stderr,
+				"%s: Cannot determine maximum number of file descriptors.\n",
+				module);
+		}
+		return max_fd;
+	}
+	for (int fd = 3; fd < max_fd; fd++) {
+		close(fd);
+	}
+	return 0;
+}
+
+/* overriding a weak function */
+void sys_arch_reboot(int type)
+{
+	char cmdline[CMDLINE_BUF_SIZE];
+	size_t len = 0;
+
+	if (read_cmdline(cmdline, &len) < 0) {
+		posix_exit(1);
+	}
+
+	char *argv[ARGV_BUF_SIZE];
+
+	if (get_args(cmdline, len, argv) < 0) {
+		posix_exit(1);
+	}
+
+	k_sched_lock();
+
+	if (close_open_fds() < 0) {
+		posix_exit(1);
+	}
+
+	fprintf(stderr, "%s: Restarting process.\n", module);
+
+	int res = execv("/proc/self/exe", argv);
```

Comment:
Just exiting this way may leave things in a bad state, it would be nice to hook the reboot logic into the exit hooks and exit.
This is getting a bit too complicated to just describe here, so I just did it in this branch:
https://github.com/aescolar/zephyr/tree/native-sim-reboot
(note I'm not done with the review, and I just tested it quickly :) )